### ISSUE TITLE
compare diff. semi-supervised learning

### ISSUE THREAD

**Issue by Edouard360** [YosefLab/scVI-dev#87]
As discussed in #59, we should consider comparing different existing semi-sup frameworks for the classification task.

A first ref is the ladder autoencoder implementation:

https://github.com/jibancanyang/Semi-supervised_Neural_Network

### On Chenlin synthetic data

- 20000 cells * 2000 genes
- 5 categories

train_labelled fraction : 0.1 => 2000 samples
train_unlabelled fraction : 0.6 => 12000 samples
validation fraction : 0.3 => 6000 samples

[NB: _validation_ is the name in _jibancanyang_ repo, but it should probably be called  _test_ ]

#### Two simple changes

In parameters change 10 and 784 (MNIST) to 5 and 2000:
    
    my_parameter = {
    ...
    'encoder_sizes': [1000, 800, 600, 400, 200, 5], # 5 = n_labels
    'decoder_sizes': [200, 400, 600, 800, 1000, 2000], # 2000 = cluster.shape[1]
    'encoder_train_bn_scaling': [True, True, True, True, True, True]
    }

In ladder.py, comment MNIST data loading out and write:

    cluster = np.load('../scVI-dev/data/cluster.npy')
    count = np.load('../scVI-dev/data/count.npy').T

    n = len(cluster)
    cluster = cluster - 1

    idx_short = np.random.permutation(n)
    cluster = cluster[idx_short]
    count = count[idx_short]

    po, pu = 0.1, 0.6 # 0.1 observed - 0.6 unobserved - 0.3 validation
    train_o_split, train_u_split = int(po * n), int((po+pu) * n)

    train_labelled_images=count[:train_o_split]
    train_labelled_labels=cluster[:train_o_split]

    train_unlabelled_images=count[train_o_split:train_u_split]
    train_unlabelled_labels=cluster[train_o_split:train_u_split]

    validation_images=count[train_u_split:]
    validation_labels=cluster[train_u_split:]

After 100 epochs, `python ladder_network/ladder.py` gives the best accuracy on validation to be 0.3038 ...

### SVAEC

On the other hand our implementation of M1+M2, with the exact same data_loaders, a pretraining of 10 epochs using a regular VAE, and then 10 more epochs using SVAEC:

    vae = VAE(gene_dataset.nb_genes, n_batch=gene_dataset.n_batches * use_batches, use_cuda=use_cuda,
              n_labels=gene_dataset.n_labels)
    train(vae, data_loader_train, data_loader_test, n_epochs=n_epochs,
          lr=lr, verbose=verbose, record_frequency=record_frequency)

    prior = torch.FloatTensor(
        [(gene_dataset.labels == i).type(torch.float32).mean() for i in range(gene_dataset.n_labels)])

    vaec = SVAEC(gene_dataset.nb_genes, n_labels=gene_dataset.n_labels, y_prior=prior, n_latent=n_latent,
                 use_cuda=use_cuda)

    vaec.z_encoder.load_state_dict(vae.z_encoder.state_dict())
    stats = train(vaec, data_loader_train, data_loader_test, n_epochs=n_epochs, lr=lr,
                  classification_ratio=alpha, verbose=verbose, record_frequency=record_frequency)

    axes[0].plot(stats.history["Accuracy_train"], label='M1+M2 (train all)')
    axes[1].plot(stats.history["Accuracy_test"])

    print("Validation : ",compute_accuracy(vaec, data_loader_validation))

Gives accuracy of: 0.72

**Comment by Edouard360** [comment]
Actually, this implementation of this [repo](https://github.com/wohlert/semi-supervised-pytorch) gives respectively, for M1+M2, applied on Chenlin dataset (same train, test, validation loaders).

# On Chenlin Dataset

## M1+M2

    # DeepGenerativeModel (M1 + M2) parameters
    y_dim = 5
    z_dim = 64
    h_dim = [256, 128, 64]

    model = DeepGenerativeModel([2000, y_dim, z_dim, h_dim])

Final accuracy on train: 1.0
Final accuracy on validation: 0.78

## Ladder

    # LadderDeepGenerativeModel parameters
    y_dim = 5
    z_dim = [64, 32, 16]
    h_dim = [256, 128, 64]
    
    model = LadderDeepGenerativeModel([2000, y_dim, z_dim, h_dim])

Final accuracy on train: 1.0
Final accuracy on validation: 0.68

<hr>

On the other hand, our current implementation of M1+M2, (which integrates the scVI decoder and therefore the modeling of the ZINB distribution), gets an equally good accuracy on validation set after 30 epochs (~0.78)

**Comment by Edouard360** [comment]
Also I am not sure about our current implementation of M1+M2, since the **LadderDeepGenerativeModel** can be represented as:
![ladderdgm](https://user-images.githubusercontent.com/15527397/39483636-e6a34cc0-4d27-11e8-84fc-25b976cb5989.png)

But in the **VFAE** paper, the generative model can be represented as:

<img width="489" alt="screen shot 2018-05-01 at 10 12 20" src="https://user-images.githubusercontent.com/15527397/39483787-624ac1d2-4d28-11e8-8251-6a73dc9d09dc.png">

And as for the encoder, both the **VFAE** paper and the original paper on Semi-supervised Learning with Deep Generative Models (Kingma 2014), agree that there shouldn't be a direct arrow from x to y (classifier), but that instead, y should be predicted from z1.

As of now, our current implementation follows this graphical representation:

![svaec-1](https://user-images.githubusercontent.com/15527397/39484006-21f758a6-4d29-11e8-8d56-17dc413aa007.png)

**Comment by jeff-regier** [comment]
What about test set accuracy? That's usually what you use to compare algorithms because both the training set and validation set were "seen" by the algorithms. (An algorithm could just "memorize" the training and validation sets, without generalizing to held out data, so any statistics about the training and validation sets aren't very useful for comparing algorithm.)
